banyand/series/trace/common_test.go

```diff
@@ -0,0 +1,270 @@
+	b.Finish(v1.MetadataEnd(b))
+	meta := new(v1.Metadata)
+	meta.Init(b.Bytes, b.Offset())
+	ts, err := traceSVC.getSeries(common.Metadata{
+		Spec: v1.GetRootAsMetadata(b.FinishedBytes(), 0),
+	})
+	assert.NoError(t, err)
+	return ts, db.GracefulStop
+}
+
+type seriesEntity struct {
+	seriesID string
+	entity   entity
+}
+
+type wantEntity struct {
+	entityID   string
+	dataBinary []byte
+	fieldsSize int
+}
+
+var gap = uint64(500 * time.Millisecond)
+
+func testData(ts uint64) []seriesEntity {
+
+	return []seriesEntity{
+		{
+			seriesID: "webapp_10.0.0.1",
+			entity: getEntityWithTS("1", []byte{11}, ts,
+				"trace_id-xxfff.111323",
+				0,
+				"webapp_id",
+				"10.0.0.1_id",
+				"/home_id",
+				"webapp",
+				"10.0.0.1",
+				"/home",
+				300,
+				1622933202000000000,
+			),
+		},
+		{
+			seriesID: "gateway_10.0.0.2",
+			entity: getEntityWithTS("2", []byte{12}, ts+gap,
+				"trace_id-xxfff.111323a",
+				1,
+			),
+		},
+		{
+			seriesID: "httpserver_10.0.0.3",
+			entity: getEntityWithTS("3", []byte{13}, ts+2*gap,
+				"trace_id-xxfff.111323",
+				1,
+				"httpserver_id",
+				"10.0.0.3_id",
+				"/home_id",
+				"httpserver",
+				"10.0.0.3",
+				"/home",
+				300,
+				1622933202000000000,
+				"GET",
+				"200",
+			),
+		},
+		{
+			seriesID: "database_10.0.0.4",
+			entity: getEntityWithTS("4", []byte{14}, ts+3*gap,
+				"trace_id-xxfff.111323",
+				0,
+				"database_id",
+				"10.0.0.4_id",
+				"/home_id",
+				"database",
+				"10.0.0.4",
+				"/home",
+				300,
+				1622933202000000000,
+				nil,
+				nil,
+				"MySQL",
+				"10.1.1.4",
+			),
+		},
+		{
+			seriesID: "mq_10.0.0.5",
+			entity: getEntityWithTS("5", []byte{15}, ts+4*gap,
+				"trace_id-zzpp.111323",
+				0,
+				"mq_id",
+				"10.0.0.5_id",
+				"/home_id",
+				"mq",
+				"10.0.0.5",
+				"/home",
+				300,
+				1622933202000000000,
+				nil,
+				nil,
+				nil,
+				nil,
+				"test_topic",
+				"10.0.0.5",
+			),
+		},
+		{
+			seriesID: "database_10.0.0.6",
+			entity: getEntityWithTS("6", []byte{16}, ts+5*gap,
+				"trace_id-zzpp.111323",
+				1,
+				"database_id",
+				"10.0.0.6_id",
+				"/home_id",
+				"database",
+				"10.0.0.6",
+				"/home",
+				300,
+				1622933202000000000,
+				nil,
+				nil,
+				"MySQL",
+				"10.1.1.6",
+			),
+		},
+		{
+			seriesID: "mq_10.0.0.7",
+			entity: getEntityWithTS("7", []byte{17}, ts+6*gap,
+				"trace_id-zzpp.111323",
+				0,
+				"nq_id",
+				"10.0.0.7_id",
+				"/home_id",
+				"mq",
+				"10.0.0.7",
+				"/home",
+				300,
+				1622933202000000000,
+				nil,
+				nil,
+				nil,
+				nil,
+				"test_topic",
+				"10.0.0.7",
+			),
+		},
+	}
+}
+
+type entity struct {
+	id     string
+	binary []byte
+	ts     uint64
+	items  []interface{}
+}
+
+func getEntity(id string, binary []byte, items ...interface{}) entity {
+	return entity{
+		id:     id,
+		binary: binary,
+		items:  items,
+	}
+}
+
+func getEntityWithTS(id string, binary []byte, ts uint64, items ...interface{}) entity {
```

Comment:
As I understood, we have to pass items in strict order since we have to keep the same order as defined in the schema, right?

Reply:
Exactly. This convention would reduce the overhead of write data.

Reply:
> Exactly. This convention would reduce the overhead of write data.

Make sense to me. But we probably need a write method per schema.